graphql/test/schema_test.ml

```diff
@@ -1,340 +1,401 @@
+        let query = "{ users { id foo } }" in
+        test_query query
+          (`Assoc
+            [
+              ( "errors",
+                `List
+                  [
+                    `Assoc
+                      [
+                        ( "message",
+                          `String "Field 'foo' is not defined on type 'user'"
+                        );
+                      ];
+                  ] );
+            ]) );
+    ( "fragments cannot form cycles",
+      `Quick,
+      fun () ->
+        let query =
+          "\n\
+          \      fragment F1 on Foo {\n\
+          \        ... on Bar {\n\
+          \          baz {\n\
+          \            ... F2\n\
+          \          }\n\
+          \        }\n\
+          \      }\n\n\
+          \      fragment F2 on Qux {\n\
+          \        ... F1\n\
+          \      }\n\n\
+          \      {\n\
+          \        ... F1\n\
+          \      }\n\
+          \    "
+        in
+        test_query query
+          (`Assoc
+            [
+              ( "errors",
+                `List
+                  [
+                    `Assoc
+                      [
+                        ("message", `String "Fragment cycle detected: F1, F2");
+                      ];
+                  ] );
+            ]) );
+    ( "fragments combine nested fields",
+      `Quick,
+      fun () ->
+        let query =
+          "\n\
+          \      query Q {\n\
+          \        users {\n\
+          \          role\n\
+          \        }\n\
+          \        ...F1\n\
+          \      }\n\
+          \      fragment F1 on query {\n\
+          \        users {\n\
+          \          name\n\
+          \        }\n\
+          \      }\n\
+          \    "
+        in
+        test_query query
+          (`Assoc
+            [
+              ( "data",
+                `Assoc
+                  [
+                    ( "users",
+                      `List
+                        [
+                          `Assoc
+                            [
+                              ("role", `String "admin");
+                              ("name", `String "Alice");
+                            ];
+                          `Assoc
+                            [
+                              ("role", `String "user");
+                              ("name", `String "Bob");
+                            ];
+                          `Assoc
+                            [
+                              ("role", `String "user");
+                              ("name", `String "Charlie");
+                            ];
+                        ] );
+                  ] );
+            ]) );
+    ( "introspection query should be accepted",
+      `Quick,
+      fun () ->
+        let query =
+          "\n\
+          \      query IntrospectionQuery {\n\
+          \        __schema {\n\
+          \          queryType { name }\n\
+          \          mutationType { name }\n\
```

Comment:
I'm not sure if there is something funky happening on these lines @andreas. I don't know this syntax (don't understand it). The tests still pass though.

Reply:
It needs to be rewritten to use `{| ... |}` rather than `" ... "` to avoid the ugly line breaks.